### ISSUE TITLE
Random characters appearing in file dialog default file name field (Windows)

### ISSUE THREAD

**Issue by AntTheAlchemist** [libsdl-org/SDL#11713]
FAO @Semphriss 🎁🎄🎁

![image](https://github.com/user-attachments/assets/10f0cec2-4668-42cd-8275-458859bac630)

**Comment by Semphriss** [comment]
Thanks, I'll give it a look! Merry Christmas to everybody who couldn't resist opening GitHub today, and everybody else as well :)

**Comment by Semphriss** [comment]
What string did you pass for the default file name?

**Comment by AntTheAlchemist** [comment]
> What string did you pass for the default file name?

I didn't 🤣.

**Comment by Semphriss** [comment]
@AntTheAlchemist How reliably can you reproduce this? I haven't succeeded yet, but it seems like the kind of bug that happens randomly and non-frequently (the file name buffer has to point to a non-zeroed memory area), which makes it quite annoying to debug.

I read the code, and I can't explain how it would happen:
```c
    filebuffer = (wchar_t *) SDL_malloc(SELECTLIST_SIZE * sizeof(wchar_t));
    SDL_memset(filebuffer, 0, SELECTLIST_SIZE * sizeof(wchar_t));

    // ...

    dialog.lpstrFile = filebuffer;
```

The only problem I can see is if `SDL_malloc` fails, then `filebuffer` would be `NULL`, but SELECTLIST_SIZE is hardcoded at 65535 and `wchar_t` is 2 bytes on Windows, so unless your memory is very fragmented, 128kb shouldn't fail to be allocated (or then SDL_memset should segfault).

I'll try with MinGW UBSan next time I get a chance, but it looks not-so-easy to debug. On another issue somebody suggested moving to `IFileDialog`, I might do that if I can't find how to fix this.

**Comment by slime73** [comment]
@Semphriss if I'm not mistaken, the code you quoted is only run in the file dialog's thread, right? Starting a thread doesn't block the main thread while that code executes, so anything referenced by that thread code which isn't guaranteed to persist (for example the default path) may be a pointer to freed memory by that point.

I just ran into this when trying out the dialog API myself. I think it applies to most of the dialog parameters since they aren't copied internally before the thread is started.

**Comment by Semphriss** [comment]
@slime73 That seems to be the case, although in this case there is no default file name, so the pointers are all `NULL`, so I'm not sure how it would happen. I'll fix the threading issues, though.

**Comment by AntTheAlchemist** [comment]
`SDL_ShowOpenFileDialog(callback, NULL, win, filters, sizeof(filters) / sizeof(*filters), "C:\\Users\\User\\Pictures\\", false);`

`default_file` gets corrupted right here. It doesn't make a difference if you pass a string literal, or a string on the heap. When the dialog thread was created, that string was okay, but when `windows_ShowFileDialog` is called, it's corrupt:

![image](https://github.com/user-attachments/assets/96769a6a-8a05-4655-ae20-383e76b45375)

**Comment by AntTheAlchemist** [comment]
This only happens in the Debug build. In the Release build, the string is okay (somehow the breakpoint I set in `windows_file_dialog_thread` still gets triggered in Release, so I was able to see).

**Comment by Semphriss** [comment]
Yes, I'm on that. I'm working somewhat more slowly these days, so I'll give priority to bugfixing over my feature-introducing PRs.

**Comment by slouken** [comment]
Thank you. We're very close to release candidate for 3.2 and features are locked, so bugs are definitely a priority.
